### PR TITLE
feat: $blobs.getUrl and $blobs.create methods

### LINKED ISSUE #104: add blob store api to mapeo project class

### PR DESCRIPTION
**Pull request by sethvincent**
closes #104 

There are a couple todos that need to be addressed that I'm not sure about. See comment below.

### CODE REVIEW

File: src/mapeo-project.js
Comment (achou11) [reply src/mapeo-project.js:132]: > TODO: where is the hostname set?

My guess is that we'd want to pass an opt to the `MapeoProject` constructor to specify the hostname. In general, I think it makes sense for the hostname to be generated by whatever is creating the `BlobServer` instance e.g. at the app level. Currently the `BlobServer` doesn't [handle multiprojects](https://github.com/digidem/mapeo-core-next/blob/52e6f12b9a16473ce6e3fcc85e716570cbfa2385/src/blob-server/index.js#L11) but that's the goal for it so that only one instance of a server is created for the entire application. 

> TODO: expose the projectId

think it has to be exposed based on [how the `BlobServer` is implemented](https://github.com/digidem/mapeo-core-next/blob/52e6f12b9a16473ce6e3fcc85e716570cbfa2385/src/blob-server/fastify-plugin.js#L60)

File: src/blob-store/index.js
Comment (achou11) [inline src/blob-store/index.js:325]: Is there any need for this to be a public method?

```suggestion
  #splitPath (filepath) {
```

File: src/blob-store/index.js
Comment (achou11) [inline src/blob-store/index.js:286]: [Client docs](https://github.com/digidem/mapeo-core-docs/blob/main/client-docs/README.md#blobcreate) specify that `metadata` is an optional arg:

```suggestion
   * @param {{ mimeType?: string }} [metadata={}]
   * @returns {Promise<{ original: Omit<BlobId, 'driveId'>, preview?: Omit<BlobId, 'driveId'>, thumbnail?: Omit<BlobId, 'driveId'> }>}
   */
  async create(filepaths, metadata = {}) {
```

**Review by achou11 — COMMENTED** [review]
feeling a little brain-drained so hopefully my comment about the hostname question makes some sense 😅

File: src/mapeo-project.js
Comment (gmaclennan) [reply src/mapeo-project.js:132]: For port we'll need to expose a method on blobServer like `getPort()` - should be async and return once the server has started (it will need to close every time the app goes into the background). We can use that to ensure that getUrl only resolves when the blob server is actually running.

**Review by gmaclennan — CHANGES_REQUESTED** [review]
Progress looks good, I made some comments.

How about keeping these methods (create() and getUrl()) in a separate class `BlobApi` that takes the blobStore as a constructor param? It feels like we're overloading the blobStore API here with something that has a different purpose.

File: src/blob-store/index.js
Comment (achou11) [reply src/blob-store/index.js:286]: as in `@param {{ mimeType: string }} metadata`?

File: src/mapeo-project.js
Comment (gmaclennan) [reply src/mapeo-project.js]: You should be able to do:

```ts
async function getPort(server: import('node:http').Server) {
  const addr = server.address()
  const port = addr && typeof addr === 'object' ? addr.port : null
  if (typeof port = 'number') return port
  await once(server, 'listening')
  return getPort(server)
}
```

You can get `server` from [`fastify.server()`](https://fastify.dev/docs/latest/Reference/Server#server)

No strong opinions on whether we pass down the fastify instance, http.server instance, or just the getPort function.

File: src/blob-api.js
Comment (gmaclennan) [inline src/blob-api.js:42]:
```suggestion
      name: hash.toString('hex'),
```

**Review by gmaclennan — CHANGES_REQUESTED** [review]
Good work on this. It needs tests though before merging, and I left a couple of comments including a suggestion about `getPort` and an undefined method `this.createWriteStream()`

File: src/blob-api.js
Comment (gmaclennan) [inline src/blob-api.js:74]: Names of all variants should be the same

File: src/blob-api.js
Comment (gmaclennan) [inline src/blob-api.js:90]: See comment above about return type

File: src/blob-api.js
Comment (gmaclennan) [inline src/blob-api.js:53]: ok to just use the random name here

File: src/mapeo-project.js
Comment (gmaclennan) [inline src/mapeo-project.js:129]: Missing blobServer param - can be added as a constructor param for MapeoProject. ping @achou we'll need to create a BlobServer instance in MapeoClient.

File: tests/helpers/blob-server.js
Comment (gmaclennan) [inline tests/helpers/blob-server.js:1]: not sure if this was intentional

File: tests/blob-api.js
Comment (gmaclennan) [inline tests/blob-api.js:31]: I think this is covered in the getUrl test below without needing to unit test getPort()?

File: tests/blob-api.js
Comment (gmaclennan) [inline tests/blob-api.js:9]: I think this is tested in the getUrl test below, which avoids needing to test the `getPort`, which is an implementation detail.

**Review by gmaclennan — CHANGES_REQUESTED** [review]
Looking good. Tests are failing though - need fixed before a merge. Plus a change needed in return type of `$blob.create`

File: src/blob-api.js
Comment (gmaclennan) [reply src/blob-api.js]: Ah very good point. It's actually a property on the writeStream e.g. `blobStore.createWriteStream().driveId` although I just realised that was not documented (my bad) and it's an awkward API too. Since the driveId for any written blobs will never change for a particular instance of the blobStore, I think it makes sense to just expose it as `blobStore.writerDriveId`, and use it from there. I've created [an issue](https://github.com/digidem/mapeo-core-next/issues/218) and [made a fix](https://github.com/digidem/mapeo-core-next/pull/219)

**Review by gmaclennan — CHANGES_REQUESTED** [review]
Great work on getting this wrapped up, looks good to merge now once the driveId question is resolved. I'll merge #219 today and then you can use `blobStore.writerDriveId` instead of `metadata.driveId`. Once that is done we can merge.

**Pull request merged**
